Objects/typeobject.c

```diff
@@ -347,23 +347,43 @@ _PyStaticType_GetBuiltins(void)
 static void
 type_set_flags(PyTypeObject *tp, unsigned long flags)
 {
-    if (tp->tp_flags & Py_TPFLAGS_READY) {
-        // It's possible the type object has been exposed to other threads
-        // if it's been marked ready.  In that case, the type lock should be
-        // held when flags are modified.
-        ASSERT_TYPE_LOCK_HELD();
-    }
-    // Since PyType_HasFeature() reads the flags without holding the type
-    // lock, we need an atomic store here.
-    FT_ATOMIC_STORE_ULONG_RELAXED(tp->tp_flags, flags);
+    // In the free-threaded build, this can only be used on types that have
+    // not yet been exposed to other threads.  Use type_set_flags_with_mask()
+    // for flags that need to be toggled afterwards.
+    tp->tp_flags = flags;
 }
 
+// Flags used by pattern matching
+#define COLLECTION_FLAGS (Py_TPFLAGS_SEQUENCE | Py_TPFLAGS_MAPPING)
+
+// This is the set of flags are allowed to be toggled after type_ready()
+// is called (and the type potentially exposed to other threads).
+#define POST_READY_FLAGS \
+    (COLLECTION_FLAGS | Py_TPFLAGS_IS_ABSTRACT | Py_TPFLAGS_HAVE_VECTORCALL)
+
+// Used to toggle type flags after the type has been created.  Only the flags
+// in POST_READY_FLAGS are allowed to be toggled.
 static void
 type_set_flags_with_mask(PyTypeObject *tp, unsigned long mask, unsigned long flags)
 {
     ASSERT_TYPE_LOCK_HELD();
+    // Non-heap types are immutable and so these flags can only be toggled
+    // after creation on heap types.
```

Comment:
I'm a bit worried about this assumption.

Reply:
Do you worry that it's not currently true or that it might be not true in a future version of Python?  The fact that non-heap types are immutable seems a pretty important assumption currently.   In `PyType_Ready()`:
```
    /* Historically, all static types were immutable. See bpo-43908 */
    if (!(type->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
        type_add_flags(type, Py_TPFLAGS_IMMUTABLETYPE);
        /* Static types must be immortal */
        _Py_SetImmortalUntracked((PyObject *)type);
    }
```

Reply:
I'm worried that it's not currently true:

* I don't really trust that code comment. Static types and C API extensions predate `Py_TPFLAGS_IMMUTABLETYPE`. I think you can support attribute modification on types if you construct your static type in the right way.
* Immutable in that context is talking about attribute modification, not changes to `tp_flags`. I don't think it covers things like `collections.abc.Iterator.register(my_type)`
* C extensions can modify type slots and tp_flags even if you can't change attributes from Python code.